### PR TITLE
Progress Reporting for template deployments

### LINKED ISSUE #987: Progress indication for commands like `create` and Long Running Operations
Didn't see an existing issue for this. Aplogies if so.

When issuing a command, such as `az vm create`, you get no indication of progress. The CLI just sits until the entire operation has completed. Looking in the portal, you can see as the storage, then networking, then VM, etc., is created. node.js CLI shows indication by default as each resource is in progress and completed, etc.

az --version
azure-cli (2016.09.05.nightly)

### LINKED ISSUE #3560: Template Progress
### Description
**Outline the issue here:**
For commands that use template deployments, the CLI can get more information about creation and success for progress.
---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here:

**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here:

**OS Version:** What OS and version are you using?  
Answer here:

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here:

### PR DESCRIPTION
**Pull request by oakeyc**
This PR adds more information for commands that are long running but use a template deployment, like 'vm create.'
fixes #3560. Fixes #987.
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [ ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by tjprescott** [comment]
I need some more context on this PR. What are the scope of the changes?  Does this just add progress indication with `az interactive` or does it change the core to report progress during normal CLI use?  What does the progress reporting look like?

**Comment by derekbekoe** [comment]
Agreed with @tjprescott. What does this look like?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 1]
```diff
@@ -31,6 +32,8 @@
 
 logger = azlogging.get_az_logger(__name__)
 
+DEFAULT_QUERY_TIME_RANGE = 3600000
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Seconds, milliseconds? Add a comment here.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 2]
```diff
@@ -137,23 +140,58 @@ def __init__(self, start_msg='', finish_msg='',
...
+    def _template_progress(self, correlation_id):
+        """ gets the progress for template deployments """
+        if correlation_id is not None:
+
+            formatter = "eventTimestamp ge {} and eventTimestamp le {}"
+
+            end_time = datetime.datetime.utcnow()
+            start_time = end_time - datetime.timedelta(seconds=DEFAULT_QUERY_TIME_RANGE)
+            odata_filters = formatter.format(start_time.strftime('%Y-%m-%dT%H:%M:%SZ'),
+                                             end_time.strftime('%Y-%m-%dT%H:%M:%SZ'))
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Is an end time needed for the query? we want everything to the current time and there could be small differences between `datetime.datetime.utcnow()` and the server data.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 3]
```diff
@@ -137,23 +140,58 @@ def __init__(self, start_msg='', finish_msg='',
...
+
+            odata_filters = "{} and {} eq '{}'".format(odata_filters, 'correlationId', correlation_id)
+
+            from azure.cli.core.commands.client_factory import get_mgmt_service_client
+            from azure.monitor import MonitorClient
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: These imports should go at the top of the function.
Import in the middle of a function isn't clean.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 4]
```diff
@@ -137,23 +140,58 @@ def __init__(self, start_msg='', finish_msg='',
...
+
+            activity_log = get_mgmt_service_client(MonitorClient).activity_logs.list(filter=odata_filters)
+            results = []
+            for index, item in enumerate(activity_log):
+                if index < 50:
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: What's the 50 for?
Looks like a magic number.
https://en.wikipedia.org/wiki/Magic_number_(programming)#Unnamed_numerical_constants

File: src/command_modules/azure-cli-interactive/HISTORY.rst [hunk 5]
```diff
@@ -6,6 +6,7 @@ Release History
 0.3.2 (unreleased)
 ++++++++++++++++++
 * --progress flag
+* Progress for Template Deployment
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-interactive/HISTORY.rst]: A release happened yesterday. Should rebase.

File: src/command_modules/azure-cli-storage/tests/recordings/test_storage_blob_no_credentials_scenario.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+interactions:
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-storage/tests/recordings/test_storage_blob_no_credentials_scenario.yaml]: Don't think this needs to be checked in.

**Comment by oakeyc** [comment]
It adds to the core of progress reporting. On the CLI, it shows when events have occurred. 
screen shot 2017-05-31 at 10 56 36 am
screen shot 2017-05-31 at 10 57 07 am

**Comment by tjprescott** [comment]
I think it would be good to demo this in scrum. My first thoughts, after looking at the screenshots:
- I don't think the zig-zagging thing will go over well. I mean, we found dots too "chatty" and they filled up the same line...
- The progress indication should be written to logger.info so that it is only visible with --verbose (arguably)
- The interleaved Accepted/Started messages are confusing and don't seem to be really telling me what I want to know, which is "what resources are created and when?".

**Comment by tjprescott** [comment]
Like we talked about in scrum, the monitor activity log likely doesn't give you any information you actually want. Instead, try looking at the `az group deployment operation list` command, which uses the following SDK method:  `azure.mgmt.resource.resources.operations.deployment_operations list`. From there you can get the status of the individual resource subdeployments and the timestamp that they completed (or failed).  This will look more like the portal when you look at the operation details for a deployment.

**Comment by oakeyc** [comment]
screen shot 2017-06-05 at 5 36 23 pm

**Comment by oakeyc** [comment]
screen shot 2017-06-06 at 3 11 59 pm

**Comment by derekbekoe** [comment]
@oakeyc please comment when this is ready for review.

**Comment by oakeyc** [comment]
@tjprescott @derekbekoe ready to review

**Comment by oakeyc** [comment]
@derekbekoe can please you take a look?

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 7]
```diff
@@ -131,26 +135,90 @@ def __init__(self, start_msg='', finish_msg='',
...
+            activity_log = get_mgmt_service_client(MonitorClient).activity_logs.list(filter=odata_filters)
+
+            results = []
+            max_events = 50  # default max value for events in list_activity_log
+            for index, item in enumerate(activity_log):
+                if index < max_events:
+                    results.append(item)
+                else:
+                    break
+                results = list(results)
```
Comment (derekbekoe) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: Isn't results already a list?

File: src/command_modules/azure-cli-interactive/HISTORY.rst [hunk 8]
```diff
@@ -6,6 +6,7 @@ Release History
 0.3.3 (2017-05-30)
 ++++++++++++++++++
 * --progress flag
+* Progress for Template Deployment
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-interactive/HISTORY.rst]: This should be in 'unreleased' section

**Comment by robdmoore** [comment]
Why is progress reporting not in the azure cli anymore? Just noticed this PR added it years back...

**Pull request merged**